src/theory/symmetry_detect.cpp

```diff
@@ -0,0 +1,576 @@
+  while (node_it != nodes.end())
+  {
+    result += " " + (*node_it).toString();
+    ++node_it;
+    if (node_it != nodes.end())
+    {
+      result += ",";
+    }
+  }
+  result += " ";
+  return result;
+}
+
+string SymmetryDetect::printNodeSet(unordered_set<Node, NodeHashFunction> nodes)
+{
+  string result;
+  unordered_set<Node, NodeHashFunction>::iterator node_it = nodes.begin();
+
+  while (node_it != nodes.end())
+  {
+    result += " " + (*node_it).toString();
+    ++node_it;
+    if (node_it != nodes.end())
+    {
+      result += ",";
+    }
+  }
+  result += " ";
+  return result;
+}
+
+void SymmetryDetect::PartitionTrie::getNewPartition(Partition& part)
+{
+  Trace("sym-dt") << "[sym-dt] Getting the new partitions from the leaves of "
+                     "the partition trie:"
+                  << std::endl;
+  if (!d_variables.empty())
+  {
+    vector<Node> vars;
+    vector<Node>::iterator var_it = d_variables.begin();
+    Node fresh_var =
+        NodeManager::currentNM()->mkSkolem("sym_bk", d_variables[0].getType());
+
+    while (var_it != d_variables.end())
+    {
+      vars.push_back(*var_it);
+      part.d_var_to_subvar[*var_it] = fresh_var;
+      ++var_it;
+    }
+    part.d_subvar_to_vars[fresh_var] = vars;
+  }
+  else
+  {
+    map<Node, PartitionTrie>::iterator part_it = d_children.begin();
+
+    while (part_it != d_children.end())
+    {
+      getNewPartition(part, part_it->second);
+      ++part_it;
+    }
+  }
+}
+
+void SymmetryDetect::PartitionTrie::getNewPartition(Partition& part,
+                                                    PartitionTrie& pt)
+{
+  if (!pt.d_variables.empty())
+  {
+    vector<Node> vars;
+    vector<Node>::iterator var_it = pt.d_variables.begin();
+    Node fresh_var = NodeManager::currentNM()->mkSkolem(
+        "sym_bk", pt.d_variables[0].getType());
+    Trace("sym-dt")
+        << "[sym-dt] A partition from leaves of the partition trie:{";
+
+    while (var_it != pt.d_variables.end())
+    {
+      Trace("sym-dt") << " " << *var_it;
+      vars.push_back(*var_it);
+      part.d_var_to_subvar[*var_it] = fresh_var;
+      ++var_it;
+    }
+    Trace("sym-dt") << " }" << endl;
+    part.d_subvar_to_vars[fresh_var] = vars;
+  }
+  else
+  {
+    map<Node, PartitionTrie>::iterator part_it = pt.d_children.begin();
+    while (part_it != pt.d_children.end())
+    {
+      getNewPartition(part, part_it->second);
+      ++part_it;
+    }
+  }
+}
+
+void SymmetryDetect::getVariables(vector<Partition>& partitions,
+                                  unordered_set<Node, NodeHashFunction>& vars)
+{
+  vector<Partition>::iterator part_it = partitions.begin();
+
+  while (part_it != partitions.end())
+  {
+    map<Node, vector<Node> >::iterator sub_var_it =
+        (*part_it).d_subvar_to_vars.begin();
+
+    while (sub_var_it != (*part_it).d_subvar_to_vars.end())
+    {
+      vector<Node>::iterator v_it = (sub_var_it->second).begin();
+
+      while (v_it != (sub_var_it->second).end())
```

Comment:
I think this loop can be replaced by `vars.insert((sub_var_it->second).begin(), (sub_var_it->second).end());`.